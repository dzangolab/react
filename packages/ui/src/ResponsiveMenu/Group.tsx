import { useState } from "react";
import { NavLink, useInRouterContext } from "react-router-dom";

import { Submenu } from "../Submenu";

import { CombinedMenuRouteType, ExtendedMenuRouteType, MenuRouteType } from ".";

interface IProperties {
  route: ExtendedMenuRouteType | MenuRouteType | any;
  displayIcon?: boolean;
  isActive?: boolean;
}

interface IMenuGroupProperties {
  routes: CombinedMenuRouteType;
  displayIcon: boolean;
}

export const MenuGroup: React.FC<IMenuGroupProperties> = ({
  routes,
  displayIcon,
}) => {
  const hasRouterContext = useInRouterContext();

  if (hasRouterContext) {
    return routes.map((route) => (
      <ResponsiveMenuItem
        key={route.name}
        route={route}
        displayIcon={displayIcon}
      />
    ));
  }

  const checkIsActive = (link: string) => {
    const pathnameArray = window.location.pathname.split("/");
    const isActive =
      window.location.pathname.startsWith(link) ||
      (pathnameArray.length && pathnameArray.includes(link));

    return isActive;
  };

  return routes.map((route) => {
    const isActive = checkIsActive(route.route || "");

    return (
      <ResponsiveMenuAnchorItem route={route} isActive={isActive || false} />
    );
  });
};

const ResponsiveMenuAnchorItem: React.FC<IProperties> = ({
  route,
  displayIcon,
  isActive,
}) => {
  const [showSubmenu, setShowSubmenu] = useState(false);

  return (
    <li key={route.name}>
      <a
        href={route.route}
        className={isActive ? "active" : undefined}
        aria-current={isActive ? "page" : undefined}
        onClick={() => route.submenu && setShowSubmenu(!showSubmenu)}
      >
        {displayIcon ? (
          <span role="icon" title={route.name}>
            {route.icon}
          </span>
        ) : null}
        <span role="label">{route.name}</span>
      </a>
      {route.submenu && <Submenu submenu={route.submenu} key={route.name} />}
    </li>
  );
};

const ResponsiveMenuItem: React.FC<IProperties> = ({ route, displayIcon }) => {
  const [showSubmenu, setShowSubmenu] = useState(false);
  return (
    <li key={route.name} aria-expanded={showSubmenu}>
      <NavLink
        to={route.route || ""}
        end={route.route === "/"}
        onClick={() => route.submenu && setShowSubmenu(!showSubmenu)}
      >
        {displayIcon ? (
          <span role="icon" title={route.name}>
            {route.icon}
          </span>
        ) : null}
        <span role="label">{route.name}</span>
      </NavLink>
      {route.submenu && <Submenu submenu={route.submenu} key={route.name} />}
    </li>
  );
};
