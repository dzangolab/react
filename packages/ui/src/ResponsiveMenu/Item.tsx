import { useState } from "react";
import { NavLink } from "react-router-dom";

import Submenu from "../Submenu";

import { ExtendedMenuRouteType } from ".";

interface IProperties {
  route: ExtendedMenuRouteType;
  displayIcon?: boolean;
}

export const ResponsiveMenuItem: React.FC<IProperties> = ({
  route,
  displayIcon,
}) => {
  const [showSubmenu, setShowSubmenu] = useState(false);

  return (
    <li key={route.name} aria-expanded={showSubmenu}>
      <NavLink
        to={route.route || ""}
        end={route.route === "/"}
        onClick={() => setShowSubmenu(!showSubmenu)}
      >
        {displayIcon ? (
          <span role="icon" title={route.name}>
            {route.icon}
          </span>
        ) : null}
        <span role="label">{route.name}</span>
      </NavLink>
      {route.submenu && <Submenu submenu={route.submenu} key={route.name} />}
    </li>
  );
};
