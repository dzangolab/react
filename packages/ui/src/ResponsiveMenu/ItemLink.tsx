import { Dispatch, SetStateAction } from "react";
import { useInRouterContext, NavLink } from "react-router-dom";

import { MenuRouteType, NestedMenuRouteType } from ".";

interface IItemLinkProperties {
  children: React.ReactNode;
  route: NestedMenuRouteType | MenuRouteType;
  setShowSubmenu: Dispatch<SetStateAction<boolean>>;
}

export const MenuItemLink = ({
  children,
  route,
  setShowSubmenu,
}: IItemLinkProperties) => {
  const hasRouterContext = useInRouterContext();

  if (hasRouterContext) {
    return (
      <NavLink
        to={route.route || ""}
        end={route.route === "/"}
        onClick={() =>
          "submenu" in route && setShowSubmenu((previous) => !previous)
        }
      >
        {children}
      </NavLink>
    );
  }

  const isActive = isNavLinkActive(route.route || "");

  return (
    <a
      href={route.route}
      className={isActive ? "active" : undefined}
      aria-current={isActive ? "page" : undefined}
      onClick={() =>
        "submenu" in route && setShowSubmenu((previous) => !previous)
      }
    >
      {children}
    </a>
  );
};

const isNavLinkActive = (link: string) => {
  {
    const pathnameArray = window.location.pathname.split("/");
    const isActive =
      window.location.pathname.startsWith(link) ||
      (pathnameArray.length && pathnameArray.includes(link));

    return isActive;
  }
};
