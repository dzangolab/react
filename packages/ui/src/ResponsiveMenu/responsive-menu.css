.responsive-menu > ul {
  flex: 1;
  flex-direction: column;
  gap: var(--grid-gap);
  list-style: none;
  padding: 0;
  margin: 0;
  display: flex;
}

.responsive-menu > ul > li {
  display: flex;
}

.responsive-menu > ul > li > a {
  padding: 1rem 0.5rem;
  color: inherit;
  text-decoration: none;
  flex: 1;
}

.responsive-menu.column > ul {
  flex-direction: column;
  gap: var(--responsive-menu-gap, 0.5rem);
  flex: 1;
}

.responsive-menu.column > ul > li {
  width: 100%;
}

.responsive-menu.column > ul > li:hover {
  background-color: #dbe9fa;
}

@media screen and (max-width: 576px) {
  .responsive-menu {
    flex: 1;
  }

  .responsive-menu > ul > li {
    display: flex;
    padding: 0;
  }

  .responsive-menu > ul > li > a {
    flex: 1;
    padding: 0.8rem 0 0.8rem 0.8rem;
    text-align: left;
  }

  .responsive-menu > ul > li :hover {
    background-color: #dbe9fa;
  }
}

@media screen and (min-width: 576px) {
  .responsive-menu > ul {
    flex-direction: row;
    gap: var(--responsive-menu-gap, 1rem);
  }
}
