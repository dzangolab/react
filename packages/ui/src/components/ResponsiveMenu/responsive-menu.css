nav.menu > ul {
    align-items: center;
    display: flex;
    flex-direction: column;
    gap: var(--grid-gap, );
    justify-content: flex-start;
    list-style: none;
    padding: 0;
  }

  nav.menu > ul > li {
    padding: 0.25rem 0;
  }

  nav.menu > ul > li > a {
    color: inherit;
    text-decoration: none;
  }

  @media screen and (min-width: 576px) {
    nav.menu > ul {
      flex-direction: row;
      gap: var(--responsive-menu-gap, 1rem);
    }
  }
