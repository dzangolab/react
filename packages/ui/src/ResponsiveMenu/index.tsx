import React from "react";
import { Link, useInRouterContext } from "react-router-dom";

import "./responsive-menu.css";

interface Properties {
  className: string;
  horizontal?: boolean;
  routes: {
    name: string;
    route: string;
  }[];
}

const ResponsiveMenu = ({
  routes,
  className,
  horizontal = false,
}: Properties) => {
  const hasRouterContext = useInRouterContext();

  if (className) {
    className = "menu" + " " + className;
  } else {
    className = "menu";
  }

  if (horizontal) {
    className += " column";
  }

  return (
    <nav className={className}>
      <ul>
        {routes.map((route, index) => (
          <li key={index}>
            {hasRouterContext ? (
              <Link to={route.route}>{route.name}</Link>
            ) : (
              <a href={route.route}>{route.name}</a>
            )}
          </li>
        ))}
      </ul>
    </nav>
  );
};

ResponsiveMenu.defaultProps = {
  className: undefined,
};

export default ResponsiveMenu;
