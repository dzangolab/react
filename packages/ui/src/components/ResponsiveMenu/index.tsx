import React from "react";
import "./responsive-menu.css";

interface Properties {
  cssClass: string;
  direction?: "column";
  routes: {
    name: string;
    route: string;
  }[];
}

const ResponsiveMenu = ({ routes, cssClass = "", direction }: Properties) => {
  return (
    <nav className={`menu ${direction} ${cssClass}`}>
      <ul>
        {routes.map((route, index) => (
          <li key={index}>
            <a href={route.route}>{route.name}</a>
          </li>
        ))}
      </ul>
    </nav>
  );
};

ResponsiveMenu.defaultProps = {
  cssClass: undefined,
};

export default ResponsiveMenu;
