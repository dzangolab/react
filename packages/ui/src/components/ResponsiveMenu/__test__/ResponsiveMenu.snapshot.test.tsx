import React from "react";
import renderer from "react-test-renderer";
import { expect, test } from "vitest";

import ResponsiveMenu from "../index";

export const menuRoutes = [
  {
    name: "Home",
    route: "/",
  },
  {
    name: "About",
    route: "/about",
  },
];

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as renderer.ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  const component = renderer.create(<ResponsiveMenu routes={menuRoutes} />);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
