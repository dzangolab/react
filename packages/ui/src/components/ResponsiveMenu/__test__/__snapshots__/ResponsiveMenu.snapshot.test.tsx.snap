// Vitest Snapshot v1

exports[`Component matches snapshot 1`] = `
<nav
  className="menu "
>
  <ul>
    <li>
      <a
        href="/"
      >
        Home
      </a>
    </li>
    <li>
      <a
        href="/about"
      >
        About
      </a>
    </li>
  </ul>
</nav>
`;
